HPCC-13251 Fix HPCC Windows VS 9 2008 Win64 errors

xwang2713 opened the pull request:
With HPCC Platform can be compiled with MS Visual Studio 9 2008 Win64.

cmake -G 'Visual Studio 9 2008 Win64' -DCMAKE_BUILD_TYPE=Release 
     -DUSE_NATIVE_LIBRARIES=OFF   -DCHECK_GIT_TAG=0 
     -DUSE_PYTHON=OFF -DUSE_V8=OFF -DUSE_JNI=OFF 
     -DUSE_RINSIDE=OFF  -DEXTERNALS_DIRECTORY=< externals_64 path > 
     -DUSE_APR=OFF -DUSE_MYSQL=OFF   -DUSE_SQLITE3=OFF -DUSE_XALAN=ON
    -DUSE_CASSANDRA=OFF -DUSE_MEMCACHED=OFF ../HPCC-Platform

To compile:   cmake --build . --config release  or build directly from Visual Studio

Internal user can download dependencies from http://10.176.32.10/data3/build/windows/external_both.

In public domain user can get it from  https://github.com/xwang2713/HPCC-Windows-Dependencies.git

@ghalliday please review

Commit:
HPCC-13251 Fix HPCC Windows VS 9 2008 Win64 errors

ghalliday commented:
Wouldn't be easier for the public repository to contain the files in the correct structure uncompressed, since git will automatically compress them, and then they could be used directly from the repository.  It would avoid any limits as well.

Also double checking are all the files open source so it is ok to distribute them?

ghalliday commented:
I would be useful if the 32 bit versions of the files were also be present in the repository.

xwang2713 commented:
OK. I will check the license and layout the as file structure. Also will add Win32 as well

ghalliday commented:
Also note, the only versions of visual studio express currently available for free are VS 2013 and  VS 2010.  Really we should also ensure we can build on these platforms.  That can be separate a separate issue though!

ghalliday commented:
Just tried that cmake command on my machine:
i) The target of the -G needed to be in dobule quotes.
ii) -DEXTERNALS_DIRECTORY= isn't right is it?
I tried
-DEXTERNALS_DIRECTORY=<location-of-externals>
but that gave complains about incompatible versions of bison.

ghalliday commented:
Ignore my comment on (ii) I added spaces to your text and now see it.

However if I set the externals to /blah/externals_64 I get a complaint about couldn't find icu.

Commit:
HPCC-13251 Update configuration for Win32

The changes will make Win32 and Win64 use the same level dependent libraries.
Remove add_definitions(/Zc:wchar_t-) which defines XMLCh as "unsigned short"

For x64 debug build need copy "C:\Program Files(x86)\Microsoft Visual Studio 9.0\VC\redist\Debug_NonRedist\amd64\Microsoft.VC90.DebugCRT\*.* to <HPCC Build>/bin/Debug/ then rebuild from Visual Studio 9 2008 (Debug x64).

Remove MTd
remove __WIN32 and reset CPPUNIT and MYSQL INCLUDE PATH
add Win64 MySQL support on Windows

ghalliday commented on rtl/eclrtl/eclinclude4.hpp:
```diff
@@ -46,7 +46,7 @@
 
 #define CHEAP_UCHAR_DEF
-#if __WIN32
+#if __WIN32 || _WIN64
```
This shouldn't be needed - _WIN32 should be defined in windows 32 and 64 bit.

However this test is wrong.  It should really be 
#ifdef _WIN32

xwang2713 commented:
I need update the some files to make build platform x86 and x64 with new libraries. Only failed one is x64 debug. I am working on HPCC 5.2.0 and 5.0.08 builds now. Will make this pull request ready tomorrow (3/26)

ghalliday commented:
I get the following problems building a 32bit system using that repository in a clean directory.  (I hit various other problems when I didn't use a clean directory...)

CMake Error: The following variables are used in this project, but they are set
to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CPPUNIT_LIBRARIES (ADVANCED)
    linked by target "datest" in directory c:/dev/hpcc/dali/datest
    linked by target "eclagent" in directory c:/dev/hpcc/ecl/eclagent
    linked by target "hql" in directory c:/dev/hpcc/ecl/hql
    linked by target "roxiemem" in directory c:/dev/hpcc/roxie/roxiemem
    linked by target "nbcd" in directory c:/dev/hpcc/rtl/nbcd
    linked by target "jhtree" in directory c:/dev/hpcc/system/jhtree
MYSQL_INCLUDE_DIR (ADVANCED)
   used as include directory in directory c:/dev/ln/esp/services/ws_attributes
   used as include directory in directory c:/dev/ln/ecl/eclrepository
   used as include directory in directory c:/dev/ln/esp/services/ws_loggingservice

-- Configuring incomplete, errors occurred!

The following helps:

Adding
`      ELSE()
        SET(CPPUNIT_INCLUDE_DIR "" )
        SET(CPPUNIT_LIBRARIES "" )
`
to commonSetup.cmake after `IF (USE_CPPUNIT)`

and `set ( MYSQL_INCLUDE_DIR "" )` for   `if(USE_MYSQL_REPOSITORY)` branch helps.

ghalliday commented:
Finally you need to change some code in the LN repository:

plugins/addrcleanlib/addrcleanlib.hpp
-#if __WIN32
should be
+#ifdef _WIN32

There are also various comments referring to __WIN32 which should be changed.
@xwang2713

xwang2713 commented:
1.  Add "ELSE() SET(CPPUNIT_INCLUDE_DIR "" ) SET(CPPUNIT_LIBRARIES "" ) "
2. Add "set ( MYSQL_INCLUDE_DIR "" )". I am not sure if I did current for this one.
3. Remove "__WIN"

ghalliday commented on cmake_modules/commonSetup.cmake:
```diff
@@ -642,6 +649,7 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
       endif(USE_OPENSSL)
 
       if(USE_MYSQL_REPOSITORY)
+        set(MYSQL_INCLUDE_DIR "")
```
This needs to be in the else branch - it will already be set up in this branch, after:

add_definitions (-D_NO_MYSQL_REPOSITORY)

ghalliday commented:
@xwang2713 Looks good - one change left.

xwang2713 commented:
Done. Correct "set(MYSQL_INCLUDE_DIR "")" location.

xwang2713 commented:
I need open a JIRA ticket for "__WIN32" against LN. Otherwise Clienttool on LN Windows will fail

xwang2713 commented:
Add MySQL Win64 support which is required for LN Windows build.
Also remove "__WIN32" in LN repository 
Disable optimization  in VS 9.2008 Win64 for some LN/plugin (otherwise generating code will run forever). Add comment in HPCC-13272 to re-visit this on Visual Studio 2013

ghalliday commented:
@xwang2713 this looks ready to take.  

However I think there is a problem with the icu libraries included in your repository.  The data dlls haven't been built correctly - which means that any use of unicode results in a runtime failure.

See http://userguide.icu-project.org/icufaq#TOC-Why-am-I-seeing-a-small-only-a-few-K-instead-of-a-large-several-megabytes-data-shared-library-icudt-Opening-ICU-services-fails-with-U_MISSING_RESOURCE_ERROR-and-u_init-returns-failure.

xwang2713 commented:
Fixed in "#if"
Rebuild ICU (include makedata). Size of icudt54.dll is much bigger than before which should include the data. Let me know if this resolve the problem or not. Thanks
